engine/crates/grafbase-engine/src/query_path.rs

```diff
@@ -0,0 +1,83 @@
+use std::fmt;
+
+use internment::ArcIntern;
+use serde::{Deserialize, Serialize};
+
+/// A path to the current location in a query
+///
+/// We use `im::Vector` to store this as we'll be making a ton of these with slight
+/// changes, and its structural sharing should make that more efficient with memory.
+///
+/// At one point this was just a reverse linked-list of references, but that was a
+/// real pain to integrate with defer & stream as the lifetimes wouldn't last long
+/// enough.
+#[derive(Debug, Default, Clone)]
+pub struct QueryPath(im::Vector<QueryPathSegment>);
+
+impl QueryPath {
+    pub fn empty() -> Self {
+        QueryPath::default()
+    }
+
+    pub fn push(&mut self, segment: impl Into<QueryPathSegment>) {
+        self.0.push_back(segment.into());
+    }
+
+    pub fn last_segment(&self) -> Option<&QueryPathSegment> {
```

Comment:
I would just use `last` and `iter`. Specifying `segment` is unnecessary IMHO, and it's not consistent with `push` either.